test/stateful_dataloader/test_state_dict.py

```diff
@@ -1054,13 +1053,145 @@ def test_json_serde_multi_process_map(self):
+                persistent_workers=pw,
+                collate_fn=identity,
+                multiprocessing_context="forkserver" if IS_MACOS else None,
+            )
+            state = dl.state_dict()
+            self.assertEqual(len(state["_snapshot"]["_worker_snapshots"]), num_workers)
+
+            exp = list(dl)
+
+            it = iter(dl)
+            for _ in range(2):
+                next(it)
+
+            dl.load_state_dict(state)
+            data = list(dl)
+            self.assertEqual(data, exp)
+
+            data2 = list(dl)
+            self.assertEqual(data2, exp)
+
+    def test_load_state_after_initial_state_dict(self):
+        for pw, interrupt in itertools.product([False, True], [2, 9]):
+            num_workers = 4
+            dataset = DummyMapDataset(100, shuffle=False)
+            dl = StatefulDataLoader(
+                dataset=dataset,
+                num_workers=num_workers,
+                persistent_workers=pw,
+                collate_fn=identity,
+                multiprocessing_context="forkserver" if IS_MACOS else None,
+            )
+
+            it = iter(dl)
+            for _ in range(interrupt):
+                next(it)
+            state = dl.state_dict()
+            exp = list(it)
+
+            dl = StatefulDataLoader(
+                dataset=dataset,
+                num_workers=num_workers,
+                persistent_workers=pw,
+                collate_fn=identity,
+                multiprocessing_context="forkserver" if IS_MACOS else None,
+            )
+            state0 = dl.state_dict()
+            self.assertEqual(len(state0["_snapshot"]["_worker_snapshots"]), num_workers)
+            dl.load_state_dict(state)
+            data = list(dl)
+            self.assertEqual(data, exp)
+
+    def test_load_state_before_initial_state_dict(self):
+        for pw, interrupt in itertools.product([False, True], [2, 9]):
+            num_workers = 4
+            dataset = DummyMapDataset(100, shuffle=False)
+            dl = StatefulDataLoader(
+                dataset=dataset,
+                num_workers=num_workers,
+                persistent_workers=pw,
+                collate_fn=identity,
+                multiprocessing_context="forkserver" if IS_MACOS else None,
+            )
+
+            it = iter(dl)
+            for _ in range(interrupt):
+                next(it)
+            state = dl.state_dict()
+            exp = list(it)
+
+            dl = StatefulDataLoader(
+                dataset=dataset,
+                num_workers=num_workers,
+                persistent_workers=pw,
+                collate_fn=identity,
+                multiprocessing_context="forkserver" if IS_MACOS else None,
+            )
+            dl.load_state_dict(state)
+            state0 = dl.state_dict()
+            self.assertEqual(state0, state)
+            data = list(dl)
+            self.assertEqual(data, exp)
+
+    def test_init_error(self):
```

Comment:
Since we're messing with the worker initialization flow, I want to make sure that the init_error and regular errors are still being handled the same way